### ISSUE TITLE
Opening Action Center sometimes disables Lightbulb

### ISSUE THREAD

**Issue by e-t-l** [Tyrrrz/LightBulb#223]
### Version

ver 2.3.3

### Details

This behavior isn't replicable 100% of the time, but frequently, when opening the Action Center (hotkey: Win + A), Lightbulb gets disabled, causing color temperature to revert to daytime level, which can be quite jarring. Closing the Action Center does not always make color temperature return to the time-appropriate level (which is why I wrote "disables" instead of "pauses"). When that happens, the easiest way to make Lightbulb resume working is to open the Lightbulb panel and click "Preview."

Note: This occurs on _both_ Windows 10 and Windows 11.

### Steps to reproduce

1. Enable Lightbulb.
2. Wait until it is a time of day that color temperature should be lower.
3. Open the Action Center using the taskbar button or by pressing Win+A.

**Comment by Tyrrrz** [comment]
I was never able to reproduce this on Windows 10. If it's happening only on Windows 11, then it could be a duplicate of #219.

**Comment by e-t-l** [comment]
Hmm odd, it was definitely happening in Win 10 too. Did you test it with any conditions for Lightbulb to be disabled (e.g. for fullscreen apps)? My suspicion was that the action center, which isn't really a traditional window, might have been interpreted as a fullscreen app, although I haven't had the chance to test this...

**Comment by e-t-l** [comment]
Could it have something to do with the setting "Default to day-time configuration" (which I have enabled)? I don't actually know exactly what that does.

**Comment by Tyrrrz** [comment]
It shouldn't. That option makes it so that your "off" configuration is your day-time configuration, instead of monitor default.

**Comment by e-t-l** [comment]
When you tested, did you have Gamma Smoothing and/or Polling enabled? (I have Smoothing ON but Polling OFF)

**Comment by Tyrrrz** [comment]
> When you tested, did you have Gamma Smoothing and/or Polling enabled? (I have Smoothing ON but Polling OFF)

Yes, I had the same

**Comment by e-t-l** [comment]
No, sorry, I was just making sure you'd tested it with the same setup as me. I'm traveling now and haven't had time to play with the configuration to try to make the bug stop occurring.

On a possibly unrelated note, I've been getting update prompts and I don't think the update is succeeding for some reason, so I may try doing a fresh reinstall and see if the bug persists..

**Comment by delneg** [comment]
Also experienced this on windows 10, version 2.4.1
Closing Action center doesn't help, only going into Lightbulb settings and adjusting the termperature does.

**Comment by e-t-l** [comment]
So, I just encountered this bug again, without involving the Action Center. I was changing my Windows keyboard layout (using the Win+Spacebar hotkey combo), which creates a flyout similar to the Action Center, so it definitely has something to do with that part of the Windows UI. Is there any way to view some sort of debug log to see what Lightbulb is doing when the bug occurs?

Or, as a workaround, is there some way to detect when screen gamma changes abruptly, and if certain conditions are met (e.g. it _should_ be nighttime mode, no Excluded Apps are on screen, Lightbulb is still running and active, etc). then Lightbulb re-sets the temperature (possibly by changing the value by 1, then changing it back)? Essentially, is there some way to make the program adjust the temperature, so that the user doesn't have to open up Lightbulb and manually adjust something in order to get it working again?

**Comment by Tyrrrz** [comment]
> Is there any way to view some sort of debug log to see what Lightbulb is doing when the bug occurs?

It most likely doesn't do anything, it's the Windows that resets gamma. It does that in certain other places as well for whatever reason.

> Or, as a workaround, is there some way to detect when screen gamma changes abruptly, and if certain conditions are met...

I wish it were that easy :( Unfortunately that's not possible.

Is this on Windows 10 or 11 by the way?

---

Side note: Microsoft wants to kill gamma control functionality in Windows -- see the note [here](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-setdevicegammaramp). They want people to use the built-in night light feature while blocking out third party developers (like myself) to provide an alternative.

**Comment by e-t-l** [comment]
Agh disgusting, Microsoft. I'm on Windows 11, but this first started back when my computer was on Win 10.

**Comment by e-t-l** [comment]
> > Or, as a workaround, is there some way to detect when screen gamma changes abruptly, and if certain conditions are met...
> 
> I wish it were that easy :( Unfortunately that's not possible

Makes sense that it wouldn't be easy to detect (although—and maybe I'm misunderstanding this—isn't that what Gamma Polling is supposed to do?)

What about the second part though? Might it be worth adding an option, say, to Lightbulb's tray context menu to re-establish the proper gamma? That way it would at least be an easy bug to ignore (whereas currently, the time it takes to open Lightbulb, go into the gamma settings, and change something manually takes 30+ seconds of looking into a high-color-temp screen, at nighttime).

**Comment by Tyrrrz** [comment]
> Makes sense that it wouldn't be easy to detect (although—and maybe I'm misunderstanding this—isn't that what Gamma Polling is supposed to do?)

Gamma polling basically just force-refreshes gamma every second. It doesn't react to events. In fact, LightBulb already subscribes to all related events that could correlate with gamma changes, but that doesn't catch everything unfortunately.

The downside of polling is that: a) continuously refreshing gamma is costly; b) doing it less frequently reduces the performance overhead but also leaves chance for larger gaps where something can reset gamma.

> What about the second part though? Might it be worth adding an option, say, to Lightbulb's tray context menu to re-establish the proper gamma? That way it would at least be an easy bug to ignore (whereas currently, the time it takes to open Lightbulb, go into the gamma settings, and change something manually takes 30+ seconds of looking into a high-color-temp screen, at nighttime).

Hmm, turning it off and on should do the trick. You can bind a global hotkey in settings and then just press it twice when you need to. Should definitely not take 30 seconds.

![image](https://user-images.githubusercontent.com/1935960/152770474-826f7488-3765-4427-b880-84057fd63a53.png)

As for me, I still can't reproduce the issue with action center on my Windows 10. It could be something that only manifests with a specific configuration of monitors or graphic devices. But even if it were possible to reproduce it, there is no guarantee there would be a way to fix it beyond just turning on gamma polling. :(

**Comment by e-t-l** [comment]
I know this repo's in maintenance mode, so I'm adding this mostly just for the sake of documenting the bug:
Disabling "pause when fullscreen" did not fix it.
Gamma polling is an effective solution. (I just turned it on so I'm hoping it doesn't impact system performance too much).

**Comment by Tyrrrz** [comment]
The latest CI build should have a workaround for this now.

**Comment by mackid1993** [comment]
This seems to still happen in Windows 11 with control center. The easiest way to repro is restart explorer and then open control center in the system tray (quick settings), Light Bulb will either switch off or flicker off and back on. f.lux isn't affected, but Iris is also as well as Sunset Screen. Night Light built into Windows is not affected.
